### PR TITLE
Rouchon1 normalisation

### PR DESCRIPTION
**Pull request by pierreguilmin**
Replaces the work of @gautierronan from #249, mostly minor modifications and reorganisation. See description there.

### CODE REVIEW

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:128]: `Trev` is upper triangular, so `Trev.mH` is lower, and this doesn't work with your function `inv_mm_invmh`.

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:140]: Same here. `T` is lower, and `inv_mm_invmh` takes an upper triangular matrix.

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:91]: These two lines require caching two objects, whereas only `T` is used in the end. Not sure how important this is, I'll let you decide whether its worth correcting.

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:15]: Probably a simple fix is to add the option `upper=True` in this function.

Comment (gautierronan) [reply dynamiqs/mesolve/rouchon.py:15]: How about `inv_kraus_matmul` for the name of this function? Not sure if it's better. I'll let you decide.

**Review by gautierronan — CHANGES_REQUESTED** [review]

File: dynamiqs/mesolve/rouchon.py
Comment (pierreguilmin) [reply dynamiqs/mesolve/rouchon.py:91]: It's not 100% ideal but I'd prefer to keep it like that for now, it makes the code clearer, and we're also going to need this `R` matrix for the SME, the solvers will look more similar. I think we can refine performance later.

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:130]:
```suggestion
        rho = kraus_map(rho, M0rev) - kraus_map(rho, self.M1srev)
```

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:59]: Also, it should be
```python
self.M0rev = cache(lambda Hnh: Srev @ M0rev)
self.M1srev = Srev @ self.M1s
```
I think in order to preserve the unit Kraus map structure.

File: dynamiqs/mesolve/rouchon.py
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:39]:
```suggestion
self.M1srev = self.M1s
```

**Review by gautierronan — CHANGES_REQUESTED** [review]

**Comment by gautierronan** [comment]
After the corrections above, we can merge I think.

**Comment by pierreguilmin** [comment]
Alright, I'll just run some quick tests for time-dep Hamiltonian to make sure it works.

**Comment by pierreguilmin** [comment]
And then I'll let you rewrite the implem for the Rouchon2 solver on top.

**Comment by gautierronan** [comment]
> Alright, I'll just run some quick tests for time-dep Hamiltonian to make sure it works.

Yes I actually wrote down that we need tests for time-dependent problems. It's a missing piece in the library right now, also useful for the adjoint dopri PR.

**Review by gautierronan — APPROVED** [review]

**Comment by pierreguilmin** [comment]
> Yes I actually wrote down that we need tests for time-dependent problems. It's a missing piece in the library right now, also useful for the adjoint dopri PR.

Cf. task 27 of the [private roadmap](https://github.com/orgs/dynamiqs/projects/2). 😉

**Pull request merged**
